src/callbacks/summarize_images.cpp

```diff
@@ -0,0 +1,369 @@
+  auto const& distdata = categorized_correctly_dist.DistData();
+  CircMat<El::Device::CPU> categorized_correctly(
+    *(distdata.grid), distdata.root);
+  El::Copy(categorized_correctly_dist, categorized_correctly);
+
+  if (categorized_correctly.Height() != El::Int(1))
+    LBANN_ERROR("Tom was wrong about this matrix. Oops.");
+
+  // Fill return value if root process
+  if (categorized_correctly.CrossRank() == categorized_correctly.Root()) {
+    // Loop over all samples -- samples are the *width* of the matrix
+    auto const num_samples = categorized_correctly.LocalWidth();
+    for (auto sample = decltype(num_samples){0}; sample < num_samples; ++sample) {
+      auto const& correctness_value = categorized_correctly.LockedMatrix()(0, sample);
+
+      if ((correctness_value != DataType(0))
+          && (correctness_value != DataType(1))) {
+        LBANN_ERROR("Invalid data from ", cat_accuracy_layer.get_name(),
+                    ". Received ", correctness_value, ", expected 0 or 1.");
+      }
+
+      if(img_indices.size() > static_cast<size_t>(num_samples) || img_counter >= m_num_images){
+        break;
+      }
+
+      if (meets_criteria(correctness_value)){
+        img_indices.push_back(std::make_pair(sample, El::Int(++img_counter)));
+      }
+    }
+  }
+
+  return img_indices;
+}
+
+bool categorical_accuracy_strategy::meets_criteria(
+  const DataType& match) const noexcept {
+  switch (m_match_type)
+  {
+  case MatchType::MATCH:
+    return (match == 1);
+  case MatchType::NOMATCH:
+    return (match == 0);
+  case MatchType::ALL:
+    return true;
+  }
+  return false;
+}
+
+std::string categorical_accuracy_strategy::get_tag(std::string const& layer_name,
+                                                   El::Int index, El::Int epoch) const {
+  // Sort by epoch
+  return build_string("epoch ", epoch, "/layer: ", layer_name,
+                      "/sample_index-", index);
+}
+
+// Builder function
+std::unique_ptr<image_output_strategy>
+build_categorical_accuracy_strategy_from_pbuf(google::protobuf::Message const& msg) {
+  using callback_type = lbann_data::Callback::CallbackSummarizeImages;
+  using strategy_type = callback_type::SelectionStrategy::CategoricalAccuracyStrategy;
+  using proto_match_type = strategy_type::MatchType;
+
+  auto ConvertToLbannType = [](proto_match_type a) {
+    return static_cast<categorical_accuracy_strategy::MatchType>(a);
+  };
+
+  const auto& strategy_msg = dynamic_cast<const strategy_type&>(msg);
+  return make_unique<categorical_accuracy_strategy>(
+    strategy_msg.accuracy_layer_name(),
+    ConvertToLbannType(strategy_msg.match_type()),
+    strategy_msg.num_images_per_epoch());
+}
+
+// End categorical_accuracy_strategy
+
+namespace {
+template <typename T>
+std::ostream& operator<<(std::ostream& os, std::vector<T> const& v)
+{
+  os << "{";
+  for (auto const& x : v)
+    os << " " << x;
+  return os << " }";
+}
+}
```

Comment:
Was this for debugging? If so, we should remove it. Otherwise, we should move it somewhere more accessible to others.